stew/io2.nim

```diff
@@ -0,0 +1,1007 @@
+  ##
+  ## Returns number of bytes written.
+  when defined(posix):
+    if len(data) > 0:
+      while true:
+        let res = write(cint(handle), unsafeAddr data[0], csize_t(len(data)))
+        if res == -1:
+          let errCode = ioLastError()
+          if errCode == EINTR:
+            continue
+          else:
+            return err(errCode)
+        else:
+          return ok(uint(res))
+    else:
+      return ok(0)
+  elif defined(windows):
+    if len(data) > 0:
+      var lpNumberOfBytesWritten = 0'u32
+      let res = writeFile(uint32(handle), unsafeAddr data[0], uint32(len(data)),
+                          lpNumberOfBytesWritten, nil)
+      if res != 0:
+        ok(lpNumberOfBytesWritten)
+      else:
+        err(ioLastError())
+    else:
+      ok(0)
+
+proc writeFile*(handle: IoHandle,
+                data: openarray[char]): IoResult[uint] {.inline.} =
+  ## Write ``data`` characters to file descriptor ``handle``.
+  ##
+  ## Returns number of characters written.
+  writeFile(handle, data.toOpenArrayByte(0, len(data) - 1))
+
+proc readFile*(handle: IoHandle,
+               data: var openarray[byte]): IoResult[uint] =
+  ## Reads ``len(data)`` bytes from file descriptor ``handle`` and store this
+  ## bytes to ``data``.
+  ##
+  ## Returns number of bytes read from file descriptor.
+  when defined(posix):
+    if len(data) > 0:
+      while true:
+        let res = read(cint(handle), unsafeAddr data[0], csize_t(len(data)))
+        if res == -1:
+          let errCode = ioLastError()
+          if errCode == EINTR:
+            continue
+          else:
+            return err(errCode)
+        else:
+          return ok(uint(res))
+    else:
+      return ok(0)
+  elif defined(windows):
+    if len(data) > 0:
+      var lpNumberOfBytesRead = 0'u32
+      let res = readFile(uint32(handle), unsafeAddr data[0], uint32(len(data)),
+                         lpNumberOfBytesRead, nil)
+      if res != 0:
+        ok(lpNumberOfBytesRead)
+      else:
+        err(ioLastError())
+    else:
+      ok(0)
+
+proc readFile*(handle: IoHandle,
+               data: var openarray[char]): IoResult[uint] {.inline.} =
+  ## Reads ``len(data)`` characters from file descriptor ``handle`` and store
+  ## this characters to ``data``.
+  ##
+  ## Returns number of bytes characters read from file descriptor.
+  readFile(handle, data.toOpenArrayByte(0, len(data) - 1))
+
+proc writeFile*(pathName: string, data: openarray[byte],
+                createMode: int = 0o644): IoResult[void] =
+  ## Opens a file named ``pathName`` for writing. Then writes the
+  ## content ``data`` completely to the file and closes the file afterwards.
+  ##
+  ## If file is not exists it will be created with permissions mask
+  ## ``createMode`` (default value is 0o644).
+  ##
+  ## If file is already exists, but file permissions are not equal to
+  ## ``createMode`` procedure will change permissions first and only after
```

Comment:
this might be annoying: if user explicitly changes permissions of an existing file and the program resets them, that's not good - I think in most cases keeping the existing permissions is a more reasonable thing to do.

Reply:
This procedure accepts path as an argument and so it do not supports append operation, it is originally designed to truncate original file and replace information in it. So its why we resetting permissions on file.